### PR TITLE
feat: Support Multiple Reactions - WPB-2974

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-2974" title="WPB-2974" target="_blank">TaskWPB-2974</a>  [iOS] Support Multiple Reactions [Add Unit Tests]
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

With this PR we add support for multiple reactions on a message. So far we could only react with 6 emojis, now we extend that support. 

#### TODO:

- [ ] Add unit tests to the logic code

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift [hunk 1]
```diff
@@ -18,111 +18,117 @@
...
+    @objc public static func addReaction(
+        _ reaction: String,
+        toMessage message: ZMConversationMessage
+    ) -> ZMClientMessage? {
+
+        guard reaction != "" else {
+            return setReactions(
+                reactions: [],
+                toMessage: message
+            )
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: Why remove all reactions if we try to set an empty reaction?

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift [hunk 2]
```diff
@@ -18,111 +18,117 @@
...
-        case "😁": return .beamingFace
-        case "☹️": return .frowningFace
-        case "👎": return .thumbsDown
-        default: return nil
+extension ZMMessage {
+
+    @discardableResult
+    @objc public static func addReaction(
+        _ reaction: String,
+        toMessage message: ZMConversationMessage
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]:
```suggestion
        to message: ZMConversationMessage
```

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift [hunk 3]
```diff
@@ -18,111 +18,117 @@
...
-        case .like: return "❤️"
-        case .thumbsUp: return "👍"
-        case .slightlySmiling: return "🙂"
-        case .beamingFace: return "😁"
-        case .frowningFace: return "☹️"
-        case .thumbsDown: return "👎"
-        }
+    @objc public static func removeReaction(
+        _ reaction: String,
+        onMessage message: ZMConversationMessage
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]:
```suggestion
        from message: ZMConversationMessage
```

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift [hunk 4]
```diff
@@ -18,111 +18,117 @@
...
-    public var displayValue: String {
-        switch unicodeValue {
-        case "❤️": return "Red Heart"
-        case "👍": return "Thumbs Up"
-        case "🙂": return "Slightly Smiling"
-        case "😁": return "Beaming Face"
-        case "☹️": return "Frowing Face"
-        case "👎": return "Thumbs Down"
-        default: return "break"
+    private static func existingReactionsBySelfUser(forMessage message: ZMConversationMessage) -> Set<String> {
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]:
```suggestion
    private static func existingReactionsBySelfUser(for message: ZMConversationMessage) -> Set<String> {
```

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift [hunk 5]
```diff
@@ -18,111 +18,117 @@
...
+
+        return Set(existingReactions)
     }
-}
 
-extension ZMMessage {
+    @discardableResult
+    static func setReactions(
+        reactions: Set<String>,
+        toMessage message: ZMConversationMessage
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]:
```suggestion
        for message: ZMConversationMessage
```

File: wire-ios-data-model/Source/Utilis/Protos/GenericMessage+Helper.swift [hunk 6]
```diff
@@ -443,8 +443,8 @@ extension Text {
 
         // We always keep the quote from the original message
         hasQuote
-            ? updatedText.quote = quote
-            : updatedText.clearQuote()
+        ? updatedText.quote = quote
+        : updatedText.clearQuote()
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Utilis/Protos/GenericMessage+Helper.swift]:
```swift
if hasQuote {
   updatedText.quote = quote
} else {
  updatedText.clearQuote()
}
```

**Review by johnxnguyen — COMMENTED** [review]
Do we have any tests for:
- setting multiple reactions
- snapshot tests for selection of multiple reactions
- reaction validation
?

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift (hunk 1 above)
Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: If I remember correctly that was a fix for this [ticket](https://wearezeta.atlassian.net/jira/software/c/projects/WPB/boards/149?modal=detail&selectedIssue=WPB-3092&assignee=6170266af6da6a006a990919).

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift (hunk 2 above)
Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift (hunk 3 above)
Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift (hunk 4 above)
Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift (hunk 5 above)
Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Source/Utilis/Protos/GenericMessage+Helper.swift (hunk 6 above)
Comment (agisilaos) [reply wire-ios-data-model/Source/Utilis/Protos/GenericMessage+Helper.swift]: Done!

**Comment by agisilaos** [comment]
> Do we have any tests for:
> 
> * setting multiple reactions
> * snapshot tests for selection of multiple reactions
> * reaction validation
>   ?

@johnxnguyen We have tests for setting multiple reactions (snapshots) and of course for setting just one reaction. I'm not sure about reaction validation. Do you think any tests are missing?

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: Wouldn't it make more sense to return nil if we try to set an empty emoji?

Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: So instead of 

```swift
return setReactions(
                reactions: [],
                toMessage: message
            )
```

to ```return nil```

Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: Done!

**Comment by johnxnguyen** [comment]
@agisilaos can you show me the snapshot tests for multiple reactions? This PR adds support for:
- the self user being able to add a new reaction to a message that they already reacted to
- the self user being able to remove a reaction from a message, and still have another reaction still there.
- receiving updates that other people reacted with 0 or more emojis for a message.

This is all new logic and it needs tests.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift [hunk 7]
```diff
@@ -37,14 +37,14 @@ extension ZMClientMessageTests_Reaction {
 
     func updateEventForAddingReaction(to message: ZMMessage, sender: ZMUser? = nil) -> ZMUpdateEvent {
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift:40]: Some test cases for setting reactions from an update event:
- given a message with no reactions, when processing the update event with one reaction, then the reaction is set on the message
- given a message with no reactions, when processing the update event with multiple reactions, then all reactions are set on the message
- given a message with reaction A, when processing the update event with reaction A, then the message still has reaction A
- given a message with reaction A, when processing the update event with reaction B, then the message only has reaction B
- given a message with reaction A, B, C, when processing the update event with reactions D, E, F, then the message has A, B, C, D, E, F
- given a message with reactions A, B, C, when processing the update event with no reactions, then the message has no reactions.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift [hunk 8]
```diff
@@ -80,13 +81,13 @@ extension ZMClientMessageTests_Reaction {
         XCTAssertTrue(waitForAllGroupsToBeEmpty(withTimeout: 0.5))
 
         XCTAssertTrue(message.cachedCategory.contains(.text))
-        XCTAssertTrue(message.cachedCategory.contains(.liked))
+        XCTAssertTrue(message.cachedCategory.contains(.reacted))
     }
 
     func testThatItDoesNOTAppendsAReactionWhenReceivingUpdateEventWithValidReaction() {
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift]: Seems like the test name should say "InvalidReaction" instead.

Comment (johnxnguyen) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift]: When testing validation, we could also have cases such as:
- given a message with no reactions, when processings an update event with reactions A, B (invalid), C, D (invalid), then the message only has reactions A and C.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift [hunk 9]
```diff
@@ -48,4 +48,75 @@ class ZMMessage_Reaction: BaseZMClientMessageTests {
...
+
+        let message = try! conversation.appendText(content: self.name) as! ZMMessage
+        message.markAsSent()
+        self.uiMOC.saveOrRollback()
+
+        XCTAssertEqual(message.deliveryState, ZMDeliveryState.sent)
+
+        message.setReactions(["😋", "😍"], forUser: selfUser)
+        XCTAssertEqual(message.selfUserReactions(), ["😋", "😍"])
+        XCTAssertEqual(message.usersReaction.count, 2)
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift]: Asserting the count is redundant due to asserting the content.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift [hunk 10]
```diff
@@ -48,4 +48,75 @@ class ZMMessage_Reaction: BaseZMClientMessageTests {
...
+
+        // WHEN
+        ZMMessage.addReaction("😎", to: message)
+        self.uiMOC.saveOrRollback()
+
+        // THEN
+        XCTAssertEqual(message.usersReaction.count, 3)
+        print("AGIS:\(message.selfUserReactions())")
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift]: Redundant assertion and also don't forget to delete the print.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift [hunk 11]
```diff
@@ -48,4 +48,75 @@ class ZMMessage_Reaction: BaseZMClientMessageTests {
...
+
+        message.setReactions(["😋", "😍"], forUser: selfUser)
+        XCTAssertEqual(message.selfUserReactions(), ["😋", "😍"])
+
+        // WHEN
+        ZMMessage.removeReaction("😋", from: message)
+        self.uiMOC.saveOrRollback()
+
+        // THEN
+        XCTAssertEqual(message.usersReaction.count, 1)
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift]: Redundant assertion.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift [hunk 12]
```diff
@@ -48,4 +48,75 @@ class ZMMessage_Reaction: BaseZMClientMessageTests {
...
+
+        message.setReactions(["😋", "😍"], forUser: selfUser)
+        XCTAssertEqual(message.selfUserReactions(), ["😋", "😍"])
+
+        // WHEN
+        ZMMessage.addReaction("", to: message)
+        self.uiMOC.saveOrRollback()
+
+        // THEN
+        XCTAssertEqual(message.usersReaction.count, 2)
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift]: Redundant reaction.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift:1]: We could also test that removing a reaction that the self user didn't react to does nothing.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessageTests.m [hunk 13]
```diff
@@ -1364,7 +1364,7 @@ - (void)testThatAddingAReactionWithUnicodeProperlyAddReactionForUserOnMessage;
     XCTAssertEqualObjects([usersThatReacted lastObject], selfUser);
 }
 
-- (void)testThatAddingAReactionWithoutUnicodeRemoveUserOnReaction;
+- (void)testThatAddingAReactionWithoutUnicodeDoesNotAddNewReactionToTheMessage;
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMMessageTests.m]: This test looks like a duplicate of one of the swift tests.

Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessageTests.m]: I was thinking the same thing but wasn't sure if I should go ahead and remove it. I will get rid of it.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift (hunk 12 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift (hunk 11 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift (hunk 10 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift (hunk 9 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift (hunk 8 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift]: Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessageTests.m (hunk 13 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessageTests.m]: Done!

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift [hunk 14]
```diff
@@ -18,111 +18,113 @@
...
     }
 
-    @discardableResult
-    @objc public static func addReaction(_ reaction: MessageReaction, toMessage message: ZMConversationMessage) -> ZMClientMessage? {
-        // confirmation that we understand the emoji
-        // the UI should never send an emoji we dont handle
-        if Reaction.transportReaction(from: reaction.unicodeValue) == .none {
-            fatal("We can't append this reaction \(reaction.unicodeValue), this is a programmer error.")
+    @objc public func setReactions(
+        _ reactions: Set<String>,
```
Comment (David-Henner) [inline wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: It took me a moment to realise the different between `reactions` and `self.reactions`. Maybe renaming this to something like `newReactions` or `updatedReactions` would make the difference more explicit?

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift [hunk 15]
```diff
@@ -18,111 +18,113 @@
...
+            }) {
+                existingReaction.mutableSetValue(forKey: ZMReactionUsersValueKey).add(user)
+            } else if Reaction.validate(unicode: reaction) {
+                let newReaction = Reaction.insertReaction(
+                    reaction,
+                    users: [user],
+                    inMessage: self
+                )
+
+                self.mutableSetValue(forKey: "reactions").add(newReaction)
```
Comment (David-Henner) [inline wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: This may be easier to read? Up to you, I don't have a strong opinion on this one 

```suggestion
        // Add all new reactions for this user.
        for reaction in reactions {
            let existingReaction = self.reactions.first(where: {
                $0.unicodeValue == reaction
            })
            
            if let existingReaction = existingReaction {
                existingReaction.mutableSetValue(forKey: ZMReactionUsersValueKey).add(user)
            } else if Reaction.validate(unicode: reaction) {
                let newReaction = Reaction.insertReaction(
                    reaction,
                    users: [user],
                    inMessage: self
                )

                mutableSetValue(forKey: "reactions").add(newReaction)
```

File: wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift [hunk 16]
```diff
@@ -19,16 +19,32 @@
...
+            let user = ZMUser.fetch(with: senderID, in: context),
             let nonce = UUID(uuidString: reaction.messageID),
-            let localMessage = ZMMessage.fetch(withNonce: nonce, for: conversation, in: moc)
+            let localMessage = ZMMessage.fetch(withNonce: nonce, for: conversation, in: context)
         else {
             return
         }
 
-        localMessage.addReaction(reaction.emoji, forUser: user)
+        localMessage.setReactions(reaction.toSet(), forUser: user)
```
Comment (David-Henner) [inline wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift]: There is something misleading in the choice of names for `WireProtos.Reaction`. It leads us to believe it's a singular reaction. But in fact it's a container for multiple reactions / emojis. 

It confuses me a little as to what a reaction is. Is a reaction a container for multiple emojis? If that's so how can an emoji also be a reaction? A reaction can't be a container for multiple reactions.

Could `WireProtos.Reaction` be renamed to something like `WireProtos.Reactions` ? Or could we call emojis `emojis` instead of `reactions` ?

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift (hunk 14 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: `updatedReactions` would be better.

Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: @David-Henner For which line are you specifically talking about?

File: wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift (hunk 16 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift]: `WireProtos.Reaction` comes from the protobuf definition, so we can't change it. Traditionally it contained only 0 or 1 reactions, but it has been repurposed to now contain 0 or more reactions. It should be called `Reactions`.

Otherwise, when there is reference to a "reaction", it generally means a single reaction, i.e a single emoji.

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift (hunk 14 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: I think he's referring to the argument `reactions: Set<String>`

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessageTests.m (hunk 13 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessageTests.m]: Did you delete this one or the swift one?

File: wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift (hunk 14 above)
Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift (hunk 8 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift]: > When testing validation, we could also have cases such as:
> 
> * given a message with no reactions, when processings an update event with reactions A, B (invalid), C, D (invalid), then the message only has reactions A and C.

@johnxnguyen I added test for that case as well. So please take a look and let me know what you think!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift (hunk 7 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift:40]: @johnxnguyen 

> given a message with reaction A, when processing the update event with reaction B, then the message only has reaction B

Is that a scenario even worth testing?

I added tests for most other cases. Please take a look and let me know what you think.

File: wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift (hunk 16 above)
Comment (David-Henner) [reply wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift]: Ok, since `WireProtos.Reaction` is not something we can easily change, could we somehow make it a little bit more explicit that we're mapping from the protobuf reaction to the message's reactions ?

Maybe something like this?

```swift
localMessage.setReactions(protobufReaction.toReactionsSet(), forUser: user)
```

Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift]: Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift:1]: Don't really get the purpose of that test?! @johnxnguyen

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift (hunk 7 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift:40]: Why not?

**Review by David-Henner — APPROVED** [review]

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift [hunk 17]
```diff
@@ -37,18 +39,38 @@ extension ZMClientMessageTests_Reaction {
...
+        // WHEN
+        performPretendingUiMocIsSyncMoc {
+            ZMClientMessage.createOrUpdate(from: event, in: self.uiMOC, prefetchResult: nil)
+        }
+        XCTAssertTrue(uiMOC.saveOrRollback())
+        XCTAssertTrue(waitForAllGroupsToBeEmpty(withTimeout: 0.5))
+
+        // THEN
+        XCTAssertEqual(message.reactions.count, 4)
+        XCTAssertEqual(message.usersReaction.count, 4)
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift]: You only need to assert `usersReaction`, and you should assert the contents and not the count.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift [hunk 18]
```diff
@@ -139,6 +182,50 @@ extension ZMClientMessageTests_Reaction {
 
         // then
         XCTAssertTrue(message.cachedCategory.contains(.text))
-        XCTAssertFalse(message.cachedCategory.contains(.liked))
+        XCTAssertFalse(message.cachedCategory.contains(.reacted))
+    }
+
+    func testThatANessageWithReactionsWhenReceivingUpdateEventWithNoReactionsRemovesAllReactions() {
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift]:
```suggestion
    func testThatAMessageWithReactionsWhenReceivingUpdateEventWithNoReactionsRemovesAllReactions() {
```

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift (hunk 7 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift:40]: I think it makes sense indeed. I will add it.

Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift:40]: @johnxnguyen, Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift
Comment (johnxnguyen) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift:1]: It's an edge case that shouldn't happen, but perhaps it could. It's a possible input hence why I suggest the test.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift [hunk 19]
```diff
@@ -139,6 +182,74 @@ extension ZMClientMessageTests_Reaction {
...
+        performPretendingUiMocIsSyncMoc {
+            ZMClientMessage.createOrUpdate(from: event, in: self.uiMOC, prefetchResult: nil)
+        }
+        XCTAssertTrue(uiMOC.saveOrRollback())
+        XCTAssertTrue(waitForAllGroupsToBeEmpty(withTimeout: 0.5))
+
+        // THEN
+        XCTAssert(message.usersReaction.contains(where: { (key: String, _) in
+            key == "🥰"
+        }))
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift]: You should be asserting who reacted too.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift (hunk 17 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift]: Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift (hunk 19 above)
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift]: Done!

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift
Comment (agisilaos) [reply wire-ios-data-model/Tests/Source/Model/Messages/ZMMessage+Reaction.swift:1]: Done!

File: wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift [hunk 20]
```diff
@@ -19,16 +19,41 @@
...
+
+    public func otherUsersReactions() -> Set<String> {
+        let result = usersReaction
+            .filter { _, users in users.contains { user in
+                !user.isSelfUser
+            }}
+            .map { $0.key }
+
+        return Set(result)
+    }
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift:57]: If this is only used in tests, then I suggest to define it as an extension in the test.

File: wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift [hunk 21]
```diff
@@ -61,14 +82,14 @@ extension ZMClientMessageTests_Reaction {
         XCTAssertTrue(uiMOC.saveOrRollback())
         XCTAssertTrue(waitForAllGroupsToBeEmpty(withTimeout: 0.5))
 
-        XCTAssertEqual(message.reactions.count, 1)
-        XCTAssertEqual(message.usersReaction.count, 1)
+        XCTAssertEqual(message.otherUsersReactions(), ["❤️"])
```
Comment (johnxnguyen) [inline wire-ios-data-model/Tests/Source/Model/Messages/ZMClientMessagesTests+Reaction.swift:85]: We really should be asserting who has which reaction, i.e user with id ABC has reactions XYZ.

**Review by johnxnguyen — APPROVED** [review]

File: wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift [hunk 22]
```diff
@@ -19,16 +19,41 @@
...
         localMessage.updateCategoryCache()
     }
+
+    func selfUserReactions() -> Set<String> {
+        let result = usersReaction
+            .filter { _, users in users.contains(where: \.isSelfUser) }
+            .map { $0.key }
+
+        return Set(result)
+    }
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift:47]: Did you remove this from the UI project? Or is there now duplication for this method?

Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift:47]: @johnxnguyen There are two different ones: One in SE and one in DataModel.

 In SE looks like:

```swift
func selfUserReactions() -> Set<Emoji> {
        let result = usersReaction
            .filter { _, users in users.contains(where: \.isSelfUser) }
            .map { Emoji(value: $0.key) }

        return Set(result)
    }
```

and in Data Model looks like:

```swift
func selfUserReactions() -> Set<String> {
        let result = usersReaction
            .filter { _, users in users.contains(where: \.isSelfUser) }
            .map { $0.key }

        return Set(result)
    }
```

Comment (johnxnguyen) [reply wire-ios-data-model/Source/Model/Message/ZMMessage+Reaction.swift:47]: It'd be good to remove this duplication. I suggest to create a ticket and address it in another PR.
